contracts/Bridge.sol

```diff
@@ -322,51 +336,43 @@ contract Bridge is Pausable, AccessControl, SafeMath {
 
         uint72 nonceAndID = (uint72(depositNonce) << 8) | uint72(chainID);
-        Proposal storage proposal = _proposals[nonceAndID][dataHash];
+        Proposal memory proposal = _proposals[nonceAndID][dataHash];
 
         require(_resourceIDToHandlerAddress[resourceID] != address(0), "no handler for resourceID");
         require(uint(proposal._status) <= 1, "proposal already passed/executed/cancelled");
-        require(!_hasVotedOnProposal[nonceAndID][dataHash][msg.sender], "relayer already voted");
-
-        if (uint(proposal._status) == 0) {
-            ++_totalProposals;
-            _proposals[nonceAndID][dataHash] = Proposal({
-                _resourceID : resourceID,
-                _dataHash : dataHash,
-                _yesVotes : new address[](1),
-                _noVotes : new address[](0),
-                _status : ProposalStatus.Active,
-                _proposedBlock : block.number
-                });
-
-            proposal._yesVotes[0] = msg.sender;
-            emit ProposalEvent(chainID, depositNonce, ProposalStatus.Active, resourceID, dataHash);
-        } else {
-            if (sub(block.number, proposal._proposedBlock) > _expiry) {
-                // if the number of blocks that has passed since this proposal was
-                // submitted exceeds the expiry threshold set, cancel the proposal
-                proposal._status = ProposalStatus.Cancelled;
-                emit ProposalEvent(chainID, depositNonce, ProposalStatus.Cancelled, resourceID, dataHash);
-            } else {
-                require(dataHash == proposal._dataHash, "datahash mismatch");
-                proposal._yesVotes.push(msg.sender);
-
-
-            }
+        require(!_hasVoted(proposal, msg.sender), "relayer already voted");
 
+        if (proposal._status == ProposalStatus.Inactive) {
+            proposal = Proposal({
+                _status : ProposalStatus.Active,
+                _yesVotes : 0,
+                _yesVotesTotal: 0,
+                _proposedBlock : uint40(block.number) // Overflow is desired.
+            });
+
+            emit ProposalEvent(chainID, depositNonce, ProposalStatus.Active, dataHash);
+        } else if (sub(block.number, proposal._proposedBlock) > _expiry) {
+            // if the number of blocks that has passed since this proposal was
+            // submitted exceeds the expiry threshold set, cancel the proposal
+            proposal._status = ProposalStatus.Cancelled;
+
+            emit ProposalEvent(chainID, depositNonce, ProposalStatus.Cancelled, dataHash);
         }
+
         if (proposal._status != ProposalStatus.Cancelled) {
-            _hasVotedOnProposal[nonceAndID][dataHash][msg.sender] = true;
-            emit ProposalVote(chainID, depositNonce, proposal._status, resourceID);
+            proposal._yesVotes = (proposal._yesVotes | _relayerBit(msg.sender)).toUint200();
+            proposal._yesVotesTotal++; // TODO: check if bit counting is cheaper.
```

Comment:
Worth creating an issue?

Reply:
Yes, I cannot quickly tell how much of a difference it will make, so will be good to investigate.